### PR TITLE
[display] Legend

### PR DESCRIPTION
**Pull request by falkecarlsen**
Create class `Legend` which can handle:
- Network name in upper left corner
- Distance legend
- Gradient-legend
- Icon-legends

The icons are scaled according to image size, although not perfectly. However, this should be good for now.

Also tries to load a pretty font, but loads default on error. ~~This probably needs a try for Windows-fonts too.~~ Fixed.

### CODE REVIEW

File: render.py [hunk 1]
```diff
@@ -40,9 +41,20 @@ def display_network(net: sumolib.net.Net, stats: ET.ElementTree, max_size: int):
     width_scale = width / city_size[0]
     height_scale = height / city_size[1]
 
+    # Load pretty font FIXME: find more ubiquitous default pretty font and maybe try for both Linux and Win
+    try:
+        font = ImageFont.truetype("/usr/share/fonts/liberation/LiberationMono-Regular.ttf", size=12)
```
Comment (jenrik) [inline render.py]: This isn't portable, it fails on systems where fonts are located elsewhere, where the font isn't installed, or it's running Windows. I suggest always using the default font, or shipping a font with releases

**Review by jenrik — DISMISSED** [review]

File: render.py [hunk 2]
```diff
@@ -17,7 +18,7 @@
 import sumolib
 
 
-def display_network(net: sumolib.net.Net, stats: ET.ElementTree, max_size: int):
+def display_network(net: sumolib.net.Net, stats: ET.ElementTree, centre, args, max_size: int):
```
Comment (NicEastvillage) [inline render.py]: You are not using `centre` anymore, and `args` could be replaced by `net_name` as that is the only thing you are using it for anyway.

Remember to update the doc string too and add type hints.

File: render.py [hunk 3]
```diff
@@ -103,19 +115,103 @@ def display_network(net: sumolib.net.Net, stats: ET.ElementTree, max_size: int):
...
-    pixels = int(meters * width_scale)
-    draw.line([2, height - 3, 2 + pixels, height - 3], (0, 0, 0), 1)
-    draw.line([2, height, 2, height - 5], (0, 0, 0), 1)
-    draw.line([2 + pixels, height, 2 + pixels, height - 5], (0, 0, 0), 1)
-    draw.text([6, height - 18], f"{meters} m", (0, 0, 0))
+    Legend(max_size, height, draw, font) \
+        .draw_network_name(args['--net-file']) \
+        .draw_distance_legend(city_size, width_scale) \
+        .draw_gradient(((0, 255, 0), (0, 0, 255)), "Pop, work gradient") \
+        .draw_legend(COLOUR_CENTRE, "Centre") \
```
Comment (NicEastvillage) [inline render.py]: Centre is not drawn atm, but appears on the legend.

**Review by NicEastvillage — CHANGES_REQUESTED** [review]
Nice legend!

The pop-work gradient not correct. It should be a two dimensional since there can be population and work places at the same time. So:
- light cyan: low pop, low work
- dark green: high pop, low work
- dark blue: low pop, high work
- dark cyan: high pop, high work
Do note that the colors are updated in #49.

Also, the distance between elements on the legend could be a bit bigger, so it is easier to tell which color the label is associated with.

**Comment by NicEastvillage** [comment]
billede
Fill in the gradient yourself

File: render.py (hunk 2 above)
Comment (falkecarlsen) [reply render.py]: Yep, it was a mistake removing centre, it's back in 0e8eefd

File: render.py (hunk 3 above)
Comment (falkecarlsen) [reply render.py]: Fixed in 0e8eefd

File: render.py (hunk 1 above)
Comment (falkecarlsen) [reply render.py]: Yep, agreed hence the FIXME-comment and the comment in the PR-description. On failing it loads the default font which work fine, but isn't very readable. 

Shipping a FOSS font might be an idea but I didn't do that, as I'd rather not further pollute the project.

File: render.py (hunk 2 above)
Comment (NicEastvillage) [reply render.py]: Why did you make it legend method??

Comment (falkecarlsen) [reply render.py]: Not sure - brainfart! Fixed in 188a4c4

Also draws the centre correctly, as it should happen before the image flipping.

Comment (falkecarlsen) [reply render.py]: Docstrings and signature fixed in d461632

File: render.py (hunk 1 above)
Comment (falkecarlsen) [reply render.py]: Windows font is added in 5959573

**Review by NicEastvillage — APPROVED** [review]

File: render.py (hunk 1 above)
Comment (jenrik) [reply render.py]: I missed the FIXME-comment.

To clarify I don't suggest shipping a font in the git repository, but tagging one on when releasing a build.

**Review by MathiasMehl — APPROVED** [review]

**Comment by falkecarlsen** [comment]
Tested scaling on max-sizes of 800, 2000, and 4000. Although they're not very pretty, they work fine.

**Pull request merged**